onnxruntime/core/providers/cuda/math/softmax_impl.cu

```diff
@@ -39,67 +39,59 @@ Status dispatch_warpwise_softmax_forward(cudaStream_t stream, output_t* dst, con
 
     // This value must match the WARP_SIZE constexpr value computed inside softmax_warp_forward.
     int warp_size = (next_power_of_two < GPU_WARP_SIZE_HOST) ? next_power_of_two : GPU_WARP_SIZE_HOST;
-
-    // This value must match the WARP_BATCH constexpr value computed inside softmax_warp_forward.
+    int threads_per_block, shared_memory_size;
+    // there are 2 options to save one row of the input matrix: register or shared memory
+    // when the number of elements is small, we use register; otherwise, we use shared memory;
     int batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
-
-    // use 128 threads per block to maximimize gpu utilization
-    constexpr int threads_per_block = 128;
-
+    if (log2_elements <= 10){
+      // This value must match the WARP_BATCH constexpr value computed inside softmax_warp_forward.
+      // use 128 threads per block to maximimize gpu utilization
+      threads_per_block = 128;
+      shared_memory_size = 0;
+    } else{
+      // 1 warp(32 threads) per block, so the index offset calculations in cuda kernel will be easier
```

Comment:
```suggestion
      // This sets there is one wrap (e.g. 32 threads) per block. The index offset calculations in this setting will be easier
      // the total block number will be equal to batch size.
```

Reply:
Is there any thing different we should take care for ROCM here?

Reply:
the accuracy should be good, as other ops use kWarpSize which is 32 also;
while rocm warp size is 64, so using 64 may have better perf;

i use USE_ROCM to keep the logic unchanged when built for RoCM, as amd folks has their optimization.